--- a/lsp-pylsp.el	2023-08-22 00:58:41.678435792 +0300
+++ b/lsp-pylsp.el	2023-08-22 00:57:26.294437440 +0300
@@ -420,7 +420,7 @@ So it will rename only references it can
 (lsp-register-client
  (make-lsp-client :new-connection (lsp-stdio-connection
                                    (lambda () lsp-pylsp-server-command))
-                  :major-modes '(python-mode cython-mode)
+                  :major-modes '(python-mode cython-mode sage-shell:sage-mode)
                   :priority -1
                   :server-id 'pylsp
                   :library-folders-fn (lambda (_workspace) lsp-clients-pylsp-library-directories)
